crates/chain/src/keychain/txout_index.rs

```diff
@@ -588,3 +588,63 @@ where
         .take_while(Result::is_ok)
         .map(Result::unwrap)
 }
+
+pub struct SpkIterator<'a> {
+    next_index: u32,
+    end: u32,
+    descriptor: Cow<'a, Descriptor<DescriptorPublicKey>>,
+    secp: Secp256k1<bitcoin::secp256k1::VerifyOnly>,
+}
+
+impl<'a> SpkIterator<'a> {
+    /// Creates a new script pubkey iterator starting at 0 from a descriptor
+    pub fn new(descriptor: Cow<'a, Descriptor<DescriptorPublicKey>>) -> Self {
+        let secp = Secp256k1::verification_only();
+        let end = if descriptor.has_wildcard() {
+            // Because we only iterate over non-hardened indexes there are 2^31 values
+            (2 ^ 31) - 1
+        } else {
+            0
+        };
+
+        Self {
+            next_index: 0,
+            end,
+            descriptor,
+            secp,
+        }
+    }
+}
+
+impl<'a> Iterator for SpkIterator<'a> {
+    type Item = (u32, Script);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // for non-wildcard descriptors:
+        // * we expect the first element to be Some((0, spk)), then will return None after.
+        // for wildcard descriptors:
+        // * we expect it to keep iterating until exhausted.
+        let has_wildcard = self.descriptor.has_wildcard();
+        if self.next_index > self.end || (has_wildcard && self.next_index > BIP32_MAX_INDEX) {
+            return None;
+        }
+
+        let script = self
+            .descriptor
+            .at_derivation_index(self.next_index)
+            .derived_descriptor(&self.secp)
+            .expect("the descriptor cannot need hardened derivation")
+            .script_pubkey();
+
+        self.next_index += 1;
+
+        Some((self.next_index, script))
+    }
+
+    fn nth(&mut self, n: usize) -> Option<Self::Item> {
+        //let n = self.next_index + n as u32;
+        //self.descriptor.at_derivation_index(n);
+        self.next_index = self.next_index.saturating_add(n as u32);
```

Comment:
I think `checking_add` is more accurate.

Reply:
In this case `saturating_add` should be OK since the `.next()` call checks that the `next_index` value isn't greater than `BIP32_MAX_INDEX`.

Reply:
Good point, however I would still prefer `checking_add` as I find it "cleaner" and easier to reason about. This has become a nit though.